stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+                          one_d out) {
+  std::vector<double> ac(a);
+  std::vector<double> bc(b);
+  Box box(ac, bc, out.result, out.err, out.kdivide);
+  return box;
+}
+
+}  // namespace internal
+
+/**
+ * Compute the n-dimensional integral of the function \f$f\f$ from \f$a\f$ to
+ \f$b\f$ within
+ * specified relative and absolute tolerances or maximum number of evaluations.
+
+ * \f$a\f$ and \f$b\f$ can be finite or infinite and should be given as vectors.
+ *
+ * The prototype for \f$f\f$ is:
+ \verbatim
+   template <typename T_x, typename T_p>
+   stan::return_type_t<T_x, T_p> f(const T_x& x, const T_p& p) {
+   using T_x_ref = stan::ref_type_t<T_x>;
+   T_x_ref x_ref = x;
+   stan::scalar_seq_view<T_x_ref> x_vec(x_ref);
+   my_params* pars = static_cast<my_params*>(p);
+   type_1 var_1 = (pars->par_1);
+   return ;
+   }
+ \endverbatim
+ *
+ * The parameters can be handed over to f as a struct:
+ \verbatim
+  struct my_params {
+  type_1 par_1;
+  type_2 par_2;
+  };
+ \endverbatim
+ *
+ * @tparam F Type of f
+ * @tparam T_pars Type of paramete-struct
+ * @tparam T_n Type of dimension
+ * @tparam T_a Type of a
+ * @tparam T_b Type of b
+ * @tparam T_maxEval Type of maximum number of evaluations
+ * @tparam T_reqAbsError Type of absolute error
+ * @tparam T_reqRelError Type of relative error
+ *
+ * @param f a functor with signature given above
+ * @param dim dimension of the integral
+ * @param a lower limit of integration as vector
+ * @param b upper limit of integration as vector
+ * @param maxEval maximal number of evaluations
+ * @param reqAbsError absolute error
+ * @param reqRelError relative error as vector
+ * @param pars parameters to be passed to f as a struct
+ * @param val correct value of integral
+ *
+ * @return The value of the n-dimensional integral of \f$f\f$ from \f$a\f$ to
+ \f$b\f$.
+ * @throw std::domain_error no errors will be thrown.
+ */
+
+// hcubature
+template <typename F, typename T_pars, typename T_n, typename T_a, typename T_b,
+          typename T_maxEval, typename T_reqAbsError, typename T_reqRelError>
+return_type_t<T_n, T_a, T_b, T_maxEval, T_reqAbsError, T_reqRelError> hcubature(
+    const F& integrand, const T_pars& pars, const T_n& n, const T_a& a,
+    const T_b& b, const T_maxEval& maxEval, const T_reqAbsError& reqAbsError,
+    const T_reqRelError& reqRelError) {
+  using T_return_type
+      = return_type_t<T_n, T_a, T_b, T_maxEval, T_reqAbsError, T_reqRelError>;
+
+  using T_a_ref = ref_type_t<T_a>;
+  using T_b_ref = ref_type_t<T_b>;
+
+  T_a_ref a_ref = a;
+  T_b_ref b_ref = b;
+
+  scalar_seq_view<T_a_ref> a_vec(a_ref);
+  scalar_seq_view<T_b_ref> b_vec(b_ref);
+
+  internal::one_d out;
+  internal::GenzMalik g;
+
+  if (n == 1) {
+    internal::gauss_kronrod(integrand, a_vec.val(0), b_vec.val(0), out, pars);
+  } else {
+    internal::make_GenzMalik(n, g);
+    internal::integrate_GenzMalik(integrand, g, n, a, b, out, pars);
+  }
```

Comment:
If `one_d` and `GenzMalik` are only used here can you just make tuples instead as the returns? imo I prefer not to have in/out parameters

struct one_d {
  double result;
  double err;
  int kdivide = 0;
};

struct GenzMalik {
  std::vector<double*> p[4];
  double w[5];
  double wd[4];
};

```suggestion
  internal::one_d out;
  internal::GenzMalik g;
  
  if (n == 1) {
    internal::gauss_kronrod(integrand, a_vec.val(0), b_vec.val(0), out, pars);
  } else {
    internal::make_GenzMalik(n, g);
    internal::integrate_GenzMalik(integrand, g, n, a, b, out, pars);
  }
```

Reply:
You mean for example for gauss_kronrod to be instead of void an 
std::tuple<double, double, int> gauss_kronrod(...)
  return std::make_tuple(result, err, kdivide)
and then in hcubature:
tie(result, err, kdivide) = gauss_kronrod(...);
?

Reply:
Ok. Next next commit.